### PR TITLE
[ONNX] Add onnx-script test cases

### LINKED ISSUE #86832: [ONNX] Add support for onnx-script as custom op
### 🚀 The feature, motivation and pitch

Currently, if a model has a **converter-unsupported, but onnx-supported** ATen Op, users
need to (1) submit a feature request, or (2) write a symbolic function to unblock the usage
of certain ONNX Ops on torchscript. However, the latter is challenge to users in terms of the complexity of writing symbolic function.

With the introduction of [onnx-script](https://github.com/microsoft/onnx-script), onnx users are able to author their own onnx functions and models using a subset of Python, which can alleviate the difficulty on writing symbolic function. Therefore, we would like to allow users to register their own onnx function as custom op to support **converter-unsupported, but onnx-supported** ATen Op.

---

## TODO

- [x] https://github.com/pytorch/pytorch/pull/86906
- [x] https://github.com/pytorch/pytorch/pull/86907
- [x] #89078 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89078
* #88376
* __->__ #86907

The test cases for #86906

### CODE REVIEW

File: test/onnx/test_onnxscript_runtime.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+        from onnxscript.onnx_opset import opset15 as op
+
+        # custom domain is needed for custom Op domain name should be
+        # aligned to the one in symbolic_fn
+        # TODO(titaiwang): make an official domain for onnxscript usage
+        custom_opset = onnxscript.values.Opset(domain="onnx-script", version=1)
+
+        @onnxscript.script(custom_opset)
+        def Selu(X, alpha: float, gamma: float):
```
Comment (gramalingam) [inline test/onnx/test_onnxscript_runtime.py:34]: This looks odd/confusin, though there is a comment below. I would suggest dropping the parameters alpha/gamma for now.

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -659,8 +659,9 @@ def _optimize_graph(
         dynamic_axes = {} if dynamic_axes is None else dynamic_axes
         _C._jit_pass_onnx_set_dynamic_input_shape(graph, dynamic_axes, input_names)
     _C._jit_pass_onnx_lint(graph)
-
+    print("torch: ", graph)
```
Comment (AllenTiTaiWang) [inline torch/onnx/utils.py:662]: I will fix this later ...

File: test/onnx/test_onnxscript_runtime.py (hunk 1 above)
Comment (AllenTiTaiWang) [reply test/onnx/test_onnxscript_runtime.py]: Done

File: test/onnx/test_onnxscript_runtime.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            gammaX = op.CastLike(gamma, X)
+            neg = gammaX * (alphaX * op.Exp(X) - alphaX)
+            pos = gammaX * X
+            zero = op.CastLike(0, X)
+            return op.Where(X <= zero, neg, pos)
+
+        def custom_selu(g: jit_utils.GraphContext, X):
+            return g.onnxscript_op(Selu, X).setType(X.type())
+
+        torch.onnx.register_custom_op_symbolic(
```
Comment (AllenTiTaiWang) [inline test/onnx/test_onnxscript_runtime.py:49]: These should be deleted

File: test/onnx/test_onnxscript_runtime.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    # 1. local function is supported after opset 15
+    # 2. onnx-script requires users to determine opset in local function
+    opset_version = 15
+
+    def test_selu_from_onnxscript_example(self):
+
+        x = torch.randn(1, 2, 3, 4, requires_grad=True)
+        model = torch.nn.SELU()
+
+        from onnxscript.onnx_opset import opset15 as op
```
Comment (BowenBao) [inline test/onnx/test_onnxscript_runtime.py:26]: Coming from https://github.com/pytorch/pytorch/pull/86906/files#r1013044361

I see this `opset15` needs to match or be compatible with the `opset_version` down in 
```python
torch.onnx.register_custom_op_symbolic(
            symbolic_name="aten::selu",
            symbolic_fn=custom_selu,
            opset_version=self.opset_version,
        )
```

Then if `opset_version` is compatible with export opset version, the onnx script function gets invoked.

So there are 3 versions. I think going forward we can try merge the first two. Read and use the opset version from the onnx_fn when registering custom op using onnx_script. We might even want to require exact opset version matching for onnxscript custom op to reduce likelihood of version issue (I'm not sure maybe @justinchuby already made the change enforcing it for all custom ops).

Comment (justinchuby) [reply test/onnx/test_onnxscript_runtime.py:26]: Currently all registered ops customs or not will follow this version resolution logic:

https://github.com/pytorch/pytorch/blob/893f8e3790df47b165025c9e6b5b37b85bdfd501/torch/onnx/_internal/registration.py#L22-L24

We can update it to create different behaviors if desired

Comment (AllenTiTaiWang) [reply test/onnx/test_onnxscript_runtime.py:26]: @justinchuby Let's have an offline meeting about this!

Comment (AllenTiTaiWang) [reply test/onnx/test_onnxscript_runtime.py:26]: One way of doing this, not elegant though, is to hide the symbolic_fn and aten op information inside onnx-script function, for example, in script decorator. In such case, we don't have to call another `register_custom_op_symbolic` for aten op, and we only need user to finish onnx-script (which makes sense they specify what aten op they would like to replace with this onnx-script function.). However, we need to find a way to parse and understand what op version we use inside the onnx function.

```python
    @onnxscript.script(custom_opset, symbolic_fn=custom_selu, aten_op="aten::selu")  # TODO: Expose used opset version 
    def Selu(X):
        alpha = 1.6732632423543772848170429916717
        gamma = 1.0507009873554804934193349852946
        alphaX = op.CastLike(alpha, X)
        gammaX = op.CastLike(gamma, X)
        neg = gammaX * (alphaX * op.Exp(X) - alphaX)
        pos = gammaX * X
        zero = op.CastLike(0, X)
        return op.Where(X <= zero, neg, pos)

    def custom_selu(g, X):
        # domain of the Op should be aligned with onnx-script
        # setType API is required for custom Op to support
        # torchscript shape type inference
        print("custom_selu is used!")
        return g.onnxscript_op(Selu, X).setType(X.type())

  #  torch.onnx.register_custom_op_symbolic(
  #      symbolic_name="aten::selu",
  #      symbolic_fn=custom_selu,
  #      opset_version=15,
  #  )
```

Comment (AllenTiTaiWang) [reply test/onnx/test_onnxscript_runtime.py:26]: In offline meeting with @justinchuby, we decided to go with this version for now, as the integration of symbolic_fn and onnx-script is in progress, we can just wait for the new design for the next release.

File: test/onnx/test_onnxscript_no_runtime.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        layer_norm_proto = onnx.load(io.BytesIO(layer_norm_onnx.getvalue()))
+
+        self.assertEqual(len(selu_proto.functions), 1)
+        self.assertEqual(len(layer_norm_proto.functions), 1)
+        self.assertEqual(selu_proto.functions[0].name, "Selu")
+        self.assertEqual(layer_norm_proto.functions[0].name, "layer_norm")
+
+    def test_loop_registration(self):
+        # Control flow is tested for _find_onnxscript_op function in torch/onnx/utils.py,
+        # which has recursive logic to go through every nodes with subgraph in model proto
```
Comment (AllenTiTaiWang) [inline test/onnx/test_onnxscript_no_runtime.py:112]: @justinchuby Added comments. LMK you have any suggestion for this comment.

Comment (justinchuby) [reply test/onnx/test_onnxscript_no_runtime.py:112]: looking good. thanks a lot!

**Review by BowenBao — APPROVED** [review]

**Pull request closed without merging**